qualtran/bloqs/data_loading/qrom.py

```diff
@@ -70,41 +74,81 @@ class QROM(UnaryIterationGate):
+        assert isinstance(self.data_or_shape, np.ndarray)
+        return self.data_or_shape
+
+    def __attrs_post_init__(self):
+        assert all([is_symbolic(s) or isinstance(s, int) for s in self.selection_bitsizes])
+        assert all([is_symbolic(t) or isinstance(t, int) for t in self.target_bitsizes])
+        assert len(self.target_bitsizes) == self.data_or_shape.shape[0], (
+            f"len(self.target_bitsizes)={len(self.target_bitsizes)} should be same as "
+            f"len(self.data)={self.data_or_shape.shape[0]}"
+        )
+        if isinstance(self.data_or_shape, np.ndarray) and not is_symbolic(*self.target_bitsizes):
+            assert all(
+                t >= int(np.max(d)).bit_length() for t, d in zip(self.target_bitsizes, self.data)
+            )
+        assert isinstance(self.selection_bitsizes, tuple)
+        assert isinstance(self.target_bitsizes, tuple)
 
     @classmethod
-    def build(cls, *data: ArrayLike, num_controls: int = 0) -> 'QROM':
-        _data = [np.array(d, dtype=int) for d in data]
-        selection_bitsizes = tuple((s - 1).bit_length() for s in _data[0].shape)
+    def build_from_data(cls, *data: ArrayLike, num_controls: SymbolicInt = 0) -> 'QROM':
+        _data = np.array([np.array(d, dtype=int) for d in data])
+        selection_bitsizes = tuple((s - 1).bit_length() for s in _data.shape[1:])
         target_bitsizes = tuple(max(int(np.max(d)).bit_length(), 1) for d in data)
         return QROM(
-            data=_data,
+            data_or_shape=_data,
             selection_bitsizes=selection_bitsizes,
             target_bitsizes=target_bitsizes,
             num_controls=num_controls,
         )
 
-    def __attrs_post_init__(self):
-        shapes = [d.shape for d in self.data]
-        assert all([isinstance(s, int) for s in self.selection_bitsizes])
-        assert all([isinstance(t, int) for t in self.target_bitsizes])
-        assert len(set(shapes)) == 1, f"Data must all have the same size: {shapes}"
-        assert len(self.target_bitsizes) == len(self.data), (
-            f"len(self.target_bitsizes)={len(self.target_bitsizes)} should be same as "
-            f"len(self.data)={len(self.data)}"
+    @classmethod
+    def build_from_bitsize(
+        cls,
+        data_len_or_shape: Union[SymbolicInt, Tuple[SymbolicInt, ...]],
+        target_bitsizes: Union[SymbolicInt, Tuple[SymbolicInt, ...]],
+        *,
+        selection_bitsizes: Tuple[SymbolicInt, ...] = (),
+        num_controls: SymbolicInt = 0,
+    ) -> 'QROM':
+        data_shape = (
+            (data_len_or_shape,) if isinstance(data_len_or_shape, int) else data_len_or_shape
         )
-        assert all(
-            t >= int(np.max(d)).bit_length() for t, d in zip(self.target_bitsizes, self.data)
+        if not isinstance(target_bitsizes, tuple):
+            target_bitsizes = (target_bitsizes,)
+        _data = Shaped((len(target_bitsizes),) + data_shape)
```

Comment:
why is len(target_bitsizes,) prepended to the shape? I was a bit confused by the notebook when it said the bloq was:
QROM(data_or_shape=Shaped(shape=(2, N, M)), selection_bitsizes=(ceiling(log2(N - 1)), ceiling(log2(M - 1))), target_bitsizes=(b1, b2), num_controls=c). Why is the shape (2, N, M)?

Reply:
I've improved the docstrings to clarify what's going on. Tl;Dr is that `data_or_shape` is now an ndarray (or shaped object) of shape `(L, S_1, S_2, ..., S_K)` where `L = number of datasets to load` and `(S_1, S_2, ..., S_K)` is the shape of each dataset to be loaded.